src/libponyrt/sched/scheduler.c

```diff
@@ -79,12 +92,38 @@ static void send_msg(uint32_t to, sched_msg_t msg, intptr_t arg)
   ponyint_messageq_push(&scheduler[to].mq, &m->msg, &m->msg);
 }
 
+static void send_msg_all_active(sched_msg_t msg, intptr_t arg)
+{
+  uint32_t current_active_scheduler_count = get_active_scheduler_count();
+  for(uint32_t i = 0; i < current_active_scheduler_count; i++)
+    send_msg(i, msg, arg);
+}
+
 static void send_msg_all(sched_msg_t msg, intptr_t arg)
 {
   for(uint32_t i = 0; i < scheduler_count; i++)
     send_msg(i, msg, arg);
 }
 
+static void wake_suspended_threads()
+{
+  uint32_t current_active_scheduler_count = get_active_scheduler_count();
+
+  // wake up any sleeping threads
+  while (current_active_scheduler_count < scheduler_count)
```

Comment:
Since this is only used at program termination, a modification here would enable the use of less expensive atomic operations while the program is running (see comment below in `steal`.)

The required change would be to set `sheduler_count_changing` to `true` before sending a signal and then wait for it to go back to `false` before locking it again and sending the next signal. This would also avoid the loop when calling `wake_suspended_threads` as it would be guaranteed that every thread has woken up when the function returns.

Reply:
Thanks for the suggestion. I've changed the logic accordingly.